crates/prune/prune/src/pruner.rs

```diff
@@ -169,81 +173,69 @@ impl<DB: Database> Pruner<DB> {
         tip_block_number: BlockNumber,
         limiter: &mut PruneLimiter,
     ) -> Result<(PrunerStats, usize, PruneProgress), PrunerError> {
-        let static_file_segments = self.static_file_segments();
-        let segments = static_file_segments
-            .iter()
-            .map(|segment| (segment, PrunePurpose::StaticFile))
-            .chain(self.segments.iter().map(|segment| (segment, PrunePurpose::User)));
-
         let mut stats = PrunerStats::new();
         let mut pruned = 0;
         let mut progress = PruneProgress::Finished;
 
-        for (segment, purpose) in segments {
-            if limiter.is_limit_reached() {
-                break
+        // Prune static file segments first
+        for segment in self.static_file_segments() {
+            let result = self.prune_segment(
+                provider,
+                tip_block_number,
+                limiter,
+                &*segment,
+                PrunePurpose::StaticFile,
+            )?;
+            match result {
+                ControlFlow::Continue(output) => {
+                    progress = output.progress;
+
+                    if output.pruned > 0 {
+                        limiter.increment_deleted_entries_count_by(output.pruned);
+                        pruned += output.pruned;
+                        stats.insert(segment.segment(), (output.progress, output.pruned));
+                    }
+                }
+                ControlFlow::LimitReached => break,
```

Comment:
static file segments will always be first, that doesn't tackle the problem in a safe way that's tolerant to all kinds of user defined prune configs. for that static files need to be flat with other segments in the ring abstraction.